tests/test_interpolation.py

```diff
@@ -753,3 +758,251 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
+        ),
+        pytest.param(
+            MissingDict(dict=SI("${identity:{0: b, 1: d}}")),
+            "dict",
+            pytest.raises(
+                InterpolationValidationError,
+                match=re.escape("Key 0 (int) is incompatible with (str)"),
+            ),
+            id="dict_key_type_mismatch",
+        ),
+    ],
+)
+def test_interpolation_type_validated_error(
+    cfg: Any,
+    key: str,
+    expected_error: Any,
+    restore_resolvers: Any,
+) -> None:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    OmegaConf.register_new_resolver("cast", cast)
+    OmegaConf.register_new_resolver("identity", lambda x: x)
+
+    cfg = OmegaConf.structured(cfg)
+
+    with expected_error:
+        cfg[key]
+
+    assert OmegaConf.select(cfg, key, throw_on_resolution_failure=False) is None
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key"),
+    [
+        pytest.param({"dict": "${identity:{a: 0, b: 1}}"}, "dict.a", id="dict"),
+        pytest.param(
+            {"dict": "${identity:{a: 0, b: {c: 1}}}"},
+            "dict.b.c",
+            id="dict_nested",
+        ),
+        pytest.param({"list": "${identity:[0, 1]}"}, "list.0", id="list"),
+        pytest.param({"list": "${identity:[0, [1, 2]]}"}, "list.1.1", id="list_nested"),
+    ],
+)
+def test_interpolation_readonly_resolver_output(
+    restore_resolvers: Any, cfg: Any, key: str
+) -> None:
+    OmegaConf.register_new_resolver("identity", lambda x: x)
+    cfg = OmegaConf.create(cfg)
+    sub_key: Any
+    parent_key, sub_key = key.rsplit(".", 1)
+    try:
+        sub_key = int(sub_key)  # convert list index to integer
+    except ValueError:
+        pass
+    parent_node = OmegaConf.select(cfg, parent_key)
+    with pytest.raises(ReadonlyConfigError):
+        parent_node[sub_key] = -1
+
+
+def test_interpolation_readonly_node() -> None:
+    cfg = OmegaConf.structured(User(name="7", age=II("name")))
+    resolved = cfg._get_node("age")._dereference_node()
+    assert resolved == 7
+    with pytest.raises(ReadonlyConfigError):
+        resolved._set_value(8)
+
+
+def test_type_validation_error_no_throw() -> None:
+    cfg = OmegaConf.structured(User(name="Bond", age=SI("${name}")))
+    bad_node = cfg._get_node("age")
+    assert bad_node._dereference_node(throw_on_resolution_failure=False) is None
+
+
+@pytest.mark.parametrize(
+    ("cfg", "expected"),
+    [
+        ({"a": 0, "b": 1}, {"a": 0, "b": 1}),
+        ({"a": "${y}"}, {"a": -1}),
+        ({"a": 0, "b": "${x.a}"}, {"a": 0, "b": 0}),
+        ({"a": 0, "b": "${.a}"}, {"a": 0, "b": 0}),
+        ({"a": "${..y}"}, {"a": -1}),
+    ],
+)
+def test_resolver_output_dictconfig(
+    restore_resolvers: Any, cfg: Dict[str, Any], expected: Dict[str, Any]
+) -> None:
+    OmegaConf.register_new_resolver("dict", lambda: cfg)
+    c = OmegaConf.create({"x": "${dict:}", "y": -1})
+    assert isinstance(c.x, DictConfig)
+    assert c.x == expected
```

Comment:
is it the same object as well?
can you add `assert c.x is expected` or `assert c.x is not expected` and add a comment about the behavior?

Reply:
Isn't it obvious? `c.x` is a DictConfig (asserted above) while `expected` is a regular Python dictionary. So they can't be the same objects. Also `expected` isn't used to compute `c.x` (only `cfg` is), so there is no way `c.x` can be the same object.

Reply:
Okay.
I was mis interpreting what you were doing.
I thought your dict function actually returned a DictConfig, and then the question is if it's returned as is being used to initialize a new one.

Reply:
Got it => better test names in 804845f